server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,650 @@
+        hidden_entries = glob.glob(os.path.join(hier.path, controller, ".*"))
+        if hidden_entries:
+            for hid_entries in hidden_entries:
+                if os.path.isfile(hid_entries):
+                    hier.add_unexpected_controllers(
+                        "unexpected_objects", controller, os.path.basename(hid_entries)
+                    )
+        controller_subdir = list()
+        for item in direct_entries:
+            base_item = os.path.basename(item)
+            if os.path.isdir(item):
+                controller_subdir.append(base_item)
+            elif os.path.islink(item):
+                symlink_item = f"{base_item} -> {os.path.realpath(item)}"
+                hier.add_unexpected_controllers(
+                    "unexpected_symlinks", controller, symlink_item
+                )
+            elif all(
+                [
+                    os.path.isfile(item),
+                    not base_item.endswith(".tar.xz"),
+                    not base_item.endswith(".tar.xz.md5"),
+                ]
+            ):
+                hier.add_unexpected_controllers(
+                    "unexpected_objects", controller, base_item
+                )
+            elif os.path.isfile(item) and (
+                base_item.endswith(".tar.xz") or base_item.endswith(".tar.xz.md5")
+            ):
+                hier.add_tarballs(controller)
+            else:
+                print(
+                    f"{base_item} item should have been handled by the above mentioned conditions. "
+                    f"It is an unexpected item which should not have occured, "
+                    f"leading to an inappropriate condition"
+                )
+        verify_subdirs(hier, controller, controller_subdir)
+        verify_prefixes(hier, controller)
+
+    return
+
+
+# Incoming Hierarchy
+def verify_tar_dirs(ihier, tarball_dirs, tblist, controller):
+    for tb in tarball_dirs:
+        if tb.endswith("unpack"):
+            tar = tb[:-7]
+            tar = f"{tar}.tar.xz"
+            val = "invalid_unpacking_dirs"
+        else:
+            tar = f"{tb}.tar.xz"
+            val = "invalid_tb_dirs"
+        tarfile = os.path.join(ihier.config.ARCHIVE, controller, tar)
+        if os.path.exists(tarfile):
+            with open(tarfile, "r") as f:
+                try:
+                    file_content = f.read(1)
+                    if file_content:
+                        continue
+                except Exception:
+                    pass
+        else:
+            tblist(val, controller, os.path.basename(tb))
+    return
+
+
+# 'ihier' is IncomingHierarchy class object
+def verify_incoming(ihier, verifylist, hierarchy=False):
+
+    for controller in verifylist:
+        ihier.add_controller(controller)
+        direct_entries = glob.iglob(
+            os.path.join(ihier.config.INCOMING, controller, "*")
+        )
+        tarball_dirs, unpacking_tarball_dirs = (list() for i in range(2))
```

Comment:
Interesting idiom; although I suspect simply tarball_dirs = list() unpacking_tarball_dirs = list() would be more efficient, as well as easier to read, I can't deny that it's fun and quirky. ;-)

Reply:
Just for grins, 3 variations from most expensive to least expensive:

```
>>> timeit.timeit("a, b=(list() for i in range(2))",number=10000000)
5.202369928359985
>>> timeit.timeit("a, b=list(),list()",number=10000000)
1.5797688961029053
>>> timeit.timeit("a=b=list()",number=10000000)
0.7393319606781006
```

Reply:
But the last one does something different from the first 2, so ignore it.

Reply:
Thanks @mnmehta @dbutenhof